File: heritage/heritage.java
import java.util.*;
import java.math.BigInteger;

class Main {
    static int N;
    static String W;
    static TreeMap<String, BigInteger> aMap, dpMap;

    static BigInteger dp(String str) {
        BigInteger count = BigInteger.valueOf(0);
        if (str.equals("")) return BigInteger.valueOf(1);
        for (int i = 1; i <= str.length(); i++) {
            String ss = "";
            String rr = "";
            for (int j = 0; j < i; j++) {
                ss += str.charAt(j);
            }
            for (int j = i; j < str.length(); j++) {
                rr += str.charAt(j);
            }
            if (aMap.containsKey(ss)) {
                if (dpMap.containsKey(rr)) count = count.add(aMap.get(ss).multiply(dpMap.get(rr)));
                else count = count.add(aMap.get(ss).multiply(dp(rr)));
            }
        }
        dpMap.put(str, count);
        return count;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();
        W = sc.next();
        aMap = new TreeMap<String, BigInteger>();
        dpMap = new TreeMap<String, BigInteger>();
        while (N > 0) {
            String str = sc.next();
            int i = sc.nextInt();
            BigInteger bi = BigInteger.valueOf(i);
            aMap.put(str, bi);
            N--;
        }
        System.out.println(dp(W).mod(BigInteger.valueOf(1000000007)));
    }
}

File: heritage/heritage.cpp
#include <iostream>
#include <string>
#include <map>

using namespace std;

int N;
string W;
map<string, long long> aMap, dpMap;

long long dp(string str) {
    long long count = 0;
    if (str == "") return 1;
    for (int i = 1; i <= str.length(); i++) {
        string ss = "";
        string rr = "";
        for (int j = 0; j < i; j++) {
            ss.push_back(str.at(j));
        }
        for (int j = i; j < str.length(); j++) {
            rr.push_back(str.at(j));
        }
        if (aMap.find(ss) != aMap.end()) {
            if (dpMap.find(rr) != dpMap.end()) count += aMap[ss] * dpMap[rr];
            else count += aMap[ss] * dp(rr);
        }
    }
    dpMap[str] = count;
    return count;
}

int main() {
    scanf("%d", &N);
    cin >> W;
    while (N--) { 
        long long x;
        string str;
        cin >> str;
        cin >> x;
        aMap[str] = x;
    }
    long long ans = dp(W);
    printf("%lld\n", ans % 1000000007);


}